include/seastar/core/iostream.hh

```diff
@@ -214,6 +230,8 @@ public:
     input_stream(input_stream&&) = default;
     input_stream& operator=(input_stream&&) = default;
     future<temporary_buffer<CharType>> read_exactly(size_t n);
+    static constexpr uint16_t DEFAULT_ALIGNMENT = alignof(void*);
+    future<temporary_buffer<CharType>> read_exactly2(size_t n, uint16_t alignment = DEFAULT_ALIGNMENT);
```

Comment:
No, if crimson-OSD supports fragmented payloads (such as SPDK), it should explicitly instruct messenger to use `ceph::net::Socket::read(size)` instead of `ceph::net::Socket::read_exactly(size, alignment)`. Because `ceph::net::Socket::read(size)` will return internally fragmented `bufferlist` as expected, and IMO it is better renamed to `ceph::net::Socket::read_fragmented(size)`.

Also, the current `ceph::net::Socket::read(size)` is already optimal for both DPDK stack and POSIX stack if OSD-side supports fragmented DATA payload:
* for DPDK: it's zero copy.
* for POSIX: it's zero copy in user-space, and also minimizes syscalls.

If OSD doesn't support fragmented payloads itself (such as kernel), `ceph::net::Socket::read_exactly(size)` still needs to be used to build up big chunks of aligned payload, regardless of whether the messenger is using Native or POSIX stack.

My point is that whether or not to use fragmented/aligned payloads should be instructed by OSD, not seastar framework. It's our (framework user) specific requirement.

Reply:
> Also, the current `ceph::net::Socket::read(size)` is already optimal for both DPDK stack and POSIX stack if OSD-side supports fragmented DATA payload:
>
>   * for DPDK: it's zero copy.
>   * for POSIX: it's zero copy in user-space, and also minimizes syscalls.

I disagree with that. For POSIX stack the SGL will be **terribly** fragmented and many syscalls will be issued because of the small, 8 KB-long prefetch buffer. For instance: reading 4 MB payload requires `4096 KB / 8 KB = 512` fragments and also 512 syscalls.

Reply:
> because of the small, 8 KB-long prefetch buffer

I think it's a separate issue, and there is already another PR addressing it (https://github.com/ceph/seastar/pull/4). My analysis shows that messenger performance is much better with larger trunks (1M), as expected. But I still don't know why `rados bench` disagreed (from kefu).